File: scripts/scrapers/education.py
import re
from .Scraper import Scraper

DATE_RANGE = re.compile(r"^([A-Za-z]{3} \d{4}) - ([A-Za-z]{3} \d{4})$")
GRADE = re.compile(r"^Grade:\s*(.+)$")


class EducationScraper(Scraper):
    def _parse_education(self, items: list[str]) -> list[dict]:
        entries = []
        i = 0
        while i < len(items):
            m = DATE_RANGE.match(items[i])
            if not m or i < 2:
                i += 1
                continue

            start, end = m.groups()
            degree = items[i - 1].strip()
            school = items[i - 2].strip()

            grade = None
            if i + 1 < len(items) and items[i + 1].startswith("Grade:"):
                grade_match = GRADE.match(items[i + 1])
                if grade_match:
                    grade = grade_match.group(1).strip()
                else:
                    grade = items[i + 1].replace("Grade:", "", 1).strip()
                i += 1

            entries.append({
                "school": school,
                "degree": degree,
                "start": start,
                "end": end,
                "grade": grade,
            })
            i += 1
        return entries
    
    def scrape(self) -> list[dict]:
        self._load_page_details("/details/education/")
        items = [s.inner_text() for s in self.page.locator("main section span[aria-hidden]").all()]
        education = self._parse_education(items)
        print(f"\nFound {len(education)} education(s):\n")
        self.page.go_back()
        return education

File: scripts/browser-scrapers/education.js
/**
 * LinkedIn Education Scraper - Run in browser console on:
 * linkedin.com/in/YOUR_USERNAME/details/education/
 *
 * Copies JSON result to clipboard. Also returns the data.
 */
(function scrapeEducation() {
  const DATE_RANGE = /^([A-Za-z]{3} \d{4}) - ([A-Za-z]{3} \d{4})$/;
  const GRADE = /^Grade:\s*(.+)$/;

  const items = [...document.querySelectorAll("main section span[aria-hidden]")].map((s) => s.innerText);
  const entries = [];
  let i = 0;

  while (i < items.length) {
    const m = items[i].match(DATE_RANGE);
    if (!m || i < 2) {
      i++;
      continue;
    }

    const [_, start, end] = m;
    const degree = items[i - 1].trim();
    const school = items[i - 2].trim();

    let grade = null;
    if (i + 1 < items.length && items[i + 1].startsWith("Grade:")) {
      const gradeMatch = items[i + 1].match(GRADE);
      grade = gradeMatch ? gradeMatch[1].trim() : items[i + 1].replace("Grade:", "", 1).trim();
      i++;
    }

    entries.push({ school, degree, start, end, grade });
    i++;
  }

  const result = JSON.stringify(entries, null, 2);
  navigator.clipboard.writeText(result);
  console.log("Education data copied to clipboard. Found", entries.length, "entries.");
  return entries;
})();
